test/ecto/query_test.exs

```diff
@@ -227,6 +227,20 @@ defmodule Ecto.QueryTest do
       assert %{post: 0} == query.aliases
     end
 
+    test "assigns a name to query source in var" do
+      posts_source = "posts"
+      query = from p in posts_source, as: :post
+
+      assert %{post: 0} == query.aliases
+    end
+
+    test "assign to source fails when query provided as source" do
+      message = ~r"can't alias query expression in `from` with `:post`"
+      assert_raise Ecto.Query.CompileError, message, fn ->
+        from p in subquery("posts"), as: :post
```

Comment:
@zoldar it feels like this should be allowed, no?

I guess the big question is: what do we want to allow and what we don't want to?

Reply:
Yeah, indeed - haven't thought that one through properly, hah. Even without the check though, moving the aliasing logic to `apply` enables applying the alias where it was previously ignored, right?